### ISSUE TITLE
DietPi-Software | Docker: Repo version bug on RPi

### ISSUE THREAD

**Issue by MichaIng** [MichaIng/DietPi#2282]
Ref:
- https://dietpi.com/phpbb/viewtopic.php?f=11&t=5227&start=10
- https://github.com/moby/moby/issues/38175#issuecomment-439733278

Not related to the fd:// issue, but a dedicated issue with most recent repo version on RPi it seems. Manual downgrading seems to work.
_________
To freeze version, until repo has the fix:
```
echo 'Package: docker-ce
Pin: version 18.06.1*
Pin-Priority: 1000' > /etc/apt/preferences.d/dietpi-docker_fix
```
Install last workking version: `G_AGI docker-ce=18.06.1~ce~3-0~raspbian`

**Note: Required on RPi only! Still unclear if it's for ARMv6 only or ARMv7 as well...
Failsafe: Pin for all RPi's!**

Stretch only, on Jessie 18.06.1 is latest offered version:
- https://download.docker.com/linux/raspbian/dists/stretch/stable/binary-armhf/Packages
- https://download.docker.com/linux/raspbian/dists/jessie/stable/binary-armhf/Packages

**Comment by MichaIng** [comment]
Alternative fix:
- Instead of freeze version to v18.06.1, better only skip faulty v18.09.0?
- This allows APT to update, as fast as a newer version is available:
```
echo 'Package: docker-ce
Pin: version 18.09.0*
Pin-Priority: -1' > /etc/apt/preferences.d/dietpi-docker_fix
```

**Comment by MichaIng** [comment]
### Test on VM Stretch:

#### After installing docker with current code:
```
root@VM-Stretch:~# dpkg -l | grep docker
ii  docker-ce                     5:18.09.0~3-0~debian-stretch   amd64        Docker: the open-source application container engine
ii  docker-ce-cli                 5:18.09.0~3-0~debian-stretch   amd64        Docker CLI: the open-source application container engine
root@VM-Stretch:~# echo 'Package: docker-ce
Pin: version 18.09.0*
Pin-Priority: -1' > /etc/apt/preferences.d/dietpi-docker_fix
```
🈴 Did not downgrade. After manual downgrade, did upgrade.

Damn, older `docker-ce` packages are in place, but not `docker-ce-cli`.

Manual downgrade of `docker-ce` works, but `docker-ce-cli` will be removed due to version incompatibility.

Old version does not depend on cli:
```
Package: docker-ce
Version: 18.03.0~ce-0~debian
Priority: optional
Section: admin
Maintainer: Docker <support@docker.com>
Installed-Size: 181 MB
Depends: iptables, init-system-helpers (>= 1.18~), libc6 (>= 2.17), libdevmapper1.02.1 (>= 2:1.02.97), libltdl7 (>= 2.4.6), libseccomp2 (>= 2.3.0), libsystemd0
Recommends: aufs-tools, ca-certificates, cgroupfs-mount | cgroup-lite, git, pigz, xz-utils
Conflicts: docker (<< 1.5~), docker-ee, docker-engine, docker-engine-cs, docker.io, lxc-docker, lxc-docker-virtual-package
Replaces: docker-engine
Homepage: https://dockerproject.org
Download-Size: 33.8 MB
APT-Sources: https://download.docker.com/linux/debian stretch/edge amd64 Packages
Description: Docker: the open-source application container engine
 Docker is an open source project to build, ship and run any application as a
 lightweight container
 .
 Docker containers are both hardware-agnostic and platform-agnostic. This means
 they can run anywhere, from your laptop to the largest EC2 compute instance and
 everything in between - and they don't require you to use a particular
 language, framework or packaging system. That makes them great building blocks
 for deploying and scaling web apps, databases, and backend services without
 depending on a particular stack or provider.
```

Ahh, new version starts with `5:`:
```
root@VM-Stretch:~# apt show -a docker-ce
Package: docker-ce
Version: 5:18.09.0~3-0~debian-stretch
Priority: optional
Section: admin
Maintainer: Docker <support@docker.com>
Installed-Size: 85.3 MB
Depends: docker-ce-cli, containerd.io, iptables, libseccomp2 (>= 2.3.0), libc6 (>= 2.17), libsystemd0
Recommends: aufs-tools, ca-certificates, cgroupfs-mount | cgroup-lite, git, pigz, xz-utils, libltdl7
Conflicts: docker (<< 1.5~), docker-engine, docker-engine-cs, docker.io, lxc-docker, lxc-docker-virtual-package
Replaces: docker-engine
Homepage: https://www.docker.com
Download-Size: 17.4 MB
APT-Sources: https://download.docker.com/linux/debian stretch/edge amd64 Packages
Description: Docker: the open-source application container engine
 Docker is a product for you to build, ship and run any application as a
 lightweight container
 .
 Docker containers are both hardware-agnostic and platform-agnostic. This means
 they can run anywhere, from your laptop to the largest cloud compute instance and
 everything in between - and they don't require you to use a particular
 language, framework or packaging system. That makes them great building blocks
 for deploying and scaling web apps, databases, and backend services without
 depending on a particular stack or provider.
```

🈯️ Works with:
```
root@VM-Stretch:~# echo 'Package: docker-ce
> Pin: version 5:18.09.0*
> Pin-Priority: -1' > /etc/apt/preferences.d/dietpi-docker_fix
root@VM-Stretch:~# cat /etc/apt/preferences.d/dietpi-docker_fix
Package: docker-ce
Pin: version 5:18.09.0*
Pin-Priority: -1
root@VM-Stretch:~# apt full-upgrade
Reading package lists... Done
Building dependency tree
Reading state information... Done
Calculating upgrade... Done
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
```

#### Recheck version string on Raspbian repo:
```
Package: docker-ce
Architecture: armhf
Version: 5:18.09.0~3-0~raspbian-stretch
```

Test if `apt upgrade` automatically downgrades the version accordingly:
```
root@VM-Stretch:~# dpkg -l | grep docker-ce
ii  docker-ce                     5:18.09.0~3-0~debian-stretch   amd64        Docker: the open-source application container engine
ii  docker-ce-cli                 5:18.09.0~3-0~debian-stretch   amd64        Docker CLI: the open-source application container engine
root@VM-Stretch:~# echo 'Package: docker-ce
> Pin: version 5:18.09.0*
> Pin-Priority: -1' > /etc/apt/preferences.d/dietpi-docker_fix
root@VM-Stretch:~# apt full-upgrade
Reading package lists... Done
Building dependency tree
Reading state information... Done
Calculating upgrade... Done
0 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.
```
🈴 Nope, manual downgrade required

🈯️ Then:
```
root@VM-Stretch:~# apt full-upgrade
Reading package lists... Done
Building dependency tree
Reading state information... Done
Calculating upgrade... Done
The following package was automatically installed and is no longer required:
  containerd.io
Use 'apt autoremove' to remove it.
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
root@VM-Stretch:~# G_AGA
[  OK  ] APT autoremove + purge, please wait...
Reading package lists...
Building dependency tree...
Reading state information...
The following packages will be REMOVED:
  containerd.io*
0 upgraded, 0 newly installed, 1 to remove and 0 not upgraded.
                                                              After this operation, 87.6 MB disk space will be freed.
(Reading database ... 16710 files and directories currently installed.)                                              (Reading database ...
Removing containerd.io (1.2.0-1) ...
(Reading database ... 16702 files and directories currently installed.)
Purging configuration files for containerd.io (1.2.0-1) ...

[  OK  ] G_AGA
root@VM-Stretch:~# dpkg -l | grep docker-ce
ii  docker-ce                     18.06.1~ce~3-0~debian          amd64        Docker: the open-source application container engine
```

Okay so this should work for RPi:
```
echo 'Package: docker-ce
Pin: version 5:18.09.0~3-0~raspbian-stretch
Pin-Priority: -1' > /etc/apt/preferences.d/dietpi-docker_fix
G_AGI docker-ce=18.06.1~ce~3-0~raspbian
G_AGA
```
Will recheck via RPi user and implement into hotfix v6.19.7
€: Being more specific in version string, in case repo update just changes the latter numbers.

**Comment by MichaIng** [comment]
Fix released with v6.19.7. Will mark as closed.
